bundles/org.openhab.binding.cbus/src/main/java/org/openhab/binding/cbus/internal/discovery/CBusGroupDiscovery.java

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.cbus.internal.discovery;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.cbus.CBusBindingConstants;
+import org.openhab.binding.cbus.handler.CBusNetworkHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.daveoxley.cbus.Application;
+import com.daveoxley.cbus.CGateException;
+import com.daveoxley.cbus.Group;
+import com.daveoxley.cbus.Network;
+
+/**
+ * The {@link CBusGroupDiscovery} class is used to discover CBus
+ * groups that are in the CBus Network
+ *
+ * @author Scott Linton - Initial contribution
+ */
+@NonNullByDefault
+public class CBusGroupDiscovery extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(CBusGroupDiscovery.class);
+
+    private final CBusNetworkHandler cbusNetworkHandler;
+    private final Semaphore scanRunning = new Semaphore(1);
+    private volatile boolean endScan;
+
+    public CBusGroupDiscovery(CBusNetworkHandler cbusNetworkHandler) {
+        super(CBusBindingConstants.SUPPORTED_THING_TYPES_UIDS, 30, false);
+        this.cbusNetworkHandler = cbusNetworkHandler;
+    }
+
+    @Override
+    protected void startScan() {
```

Comment:
If you want to make sure that only one thread runs this at a time you can just make it synchronized.
```suggestion
    protected synchronized void startScan() {
```

Reply:
updated